includes/Checker/Checks/Trademarks_Check.php

```diff
@@ -0,0 +1,543 @@
+
+	/**
+	 * Bitwise flags to control check behavior.
+	 *
+	 * @since n.e.x.t
+	 * @var int
+	 */
+	protected $flags = 0;
+
+	/**
+	 * Constructor.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param int $flags Bitwise flags to control check behavior.
+	 */
+	public function __construct( $flags = self::TYPE_ALL ) {
+		$this->flags = $flags;
+	}
+
+	/**
+	 * Gets the categories for the check.
+	 *
+	 * Every check must have at least one category.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @return array The categories for the check.
+	 */
+	public function get_categories() {
+		return array( Check_Categories::CATEGORY_PLUGIN_REPO );
+	}
+
+	/**
+	 * Check for trademarks.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param Check_Result $result The Check Result to amend.
+	 * @param array        $files  Array of plugin files.
+	 */
+	protected function check_files( Check_Result $result, array $files ) {
+
+		// Check the trademarks in readme file plugin name.
+		if ( $this->flags & self::TYPE_README ) {
+			$this->check_for_name_in_readme( $result, $files );
+		}
+
+		// Check the trademarks in plugin name.
+		if ( $this->flags & self::TYPE_NAME ) {
+			$this->check_for_name_in_main_file( $result );
+		}
+
+		// Check the trademarks in plugin slug.
+		if ( $this->flags & self::TYPE_SLUG ) {
+			$this->check_for_slug( $result );
+		}
+	}
+
+	/**
+	 * Checks the trademarks in readme file plugin name.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param Check_Result $result The Check Result to amend.
+	 * @param array        $files  Array of plugin files.
+	 */
+	private function check_for_name_in_readme( Check_Result $result, array $files ) {
+		$plugin_relative_path = $result->plugin()->path();
+
+		// Find the readme file.
+		$readme_list = self::filter_files_by_regex( $files, '/readme\.(txt|md)$/i' );
+
+		// Filter the readme files located at root.
+		$potential_readme_files = array_filter(
+			$readme_list,
+			function ( $file ) use ( $plugin_relative_path ) {
+				$file = str_replace( $plugin_relative_path, '', $file );
+				if ( ! strpos( $file, '/' ) ) {
+					return true;
+				}
+			}
+		);
+
+		// Find the .txt versions of the readme files.
+		$readme_txt = array_filter(
+			$potential_readme_files,
+			function ( $file ) {
+				return preg_match( '/^readme\.txt$/i', basename( $file ) );
+			}
+		);
+
+		// Find the .md versions of the readme files.
+		$readme_md = array_filter(
+			$potential_readme_files,
+			function ( $file ) {
+				return preg_match( '/^readme\.md$/i', basename( $file ) );
+			}
+		);
+
+		// If there's a .txt version, ignore .md versions.
+		$readme = ( ! empty( $readme_txt ) ) ? $readme_txt : $readme_md;
+
+		// If the readme file does not exist, then skip test.
+		if ( empty( $readme ) ) {
+			return;
+		}
+
+		$matches = array();
+		// Get the plugin name from readme file.
+		$file = self::file_preg_match( '/===(.*)===/i', $files, $matches );
+
+		if ( ! $file ) {
+			return;
+		}
+
+		$name = isset( $matches[1] ) ? $matches[1] : '';
```

Comment:
This check strictly isn't needed, since `$matches[1]` should always be set unless `! $file`, in which case this logic isn't reached. But still +1 for defensive coding here. My only suggestion then would be to bail early in that case too, since an empty name wouldn't make sense to check.
```suggestion
		if ( ! $file || ! isset( $matches[1] ) ) {
			return;
		}

		$name = $matches[1];
```